main/src/library/List.flix

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2016 Liam Palmer
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace List {
+
+    /**
+     * Returns the number of elements in `xs`.
+     */
+    def size[a](xs: List[a]): Int32 = match xs with {
+        case Nil => 0
+        case y :: ys => size(ys) + 1
+    }
+
+    /**
+     * Returns the first element of `xs`.
+     * Throws an exception if `xs` is empty.
+     */
+    def head[a](xs: List[a]): a = match xs with {
+        case Nil => ???
```

Comment:
We don't actually have exceptions in Flix, but I can see that `???` is close to one. The code itself is fine, but I think we should write the documentation in the style of: `Returns the first element of `xs` which must be non-empty`. Although it makes sense that calling head with an empty list gives an exception, it is not so clear-cut, since evaluation of Flix is non-deterministic, so you can get into situations where you sometimes  get an exception, and other times not. So I think we should (at least not yet) use the word exception and instead just state the requirement "... which must be non-empty" and similar phrasings.